File: Beginner/Python/SortAndSearch/FirstBadVersion.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
# def isBadVersion(version):

class Solution(object):
    def firstBadVersion(self, n):
        """
            :type n: int
            :rtype: int
            """
        # Timeout method
        # fast = slow = n
        # while isBadVersion(fast) and isBadVersion(slow):
        #     fast -= 2
        #     slow -= 1
        # while not isBadVersion(fast) and isBadVersion(slow):
        #     fast -= 1
        #     slow -= 1
        # return fast if isBadVersion(fast) else slow + 1
        
        # Find the first True
        class Wrap:
            def __getitem__(self, i):
                return isBadVersion(i)
        return bisect.bisect(Wrap(), False, 0, n)

File: Beginner/Java/SortingAndSearching/FirstBadVersion.java
/* The isBadVersion API is defined in the parent class VersionControl.
 boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        int start = 1, end = n;
        while (start < end) {
            int mid = start + (end - start) / 2;
            if (!isBadVersion(mid))
                start = mid + 1;
            else
                end = mid;
        }
        return start;
    }
    
    //     Stack Overflow
    // public int checker(int first, int last) {
    //     int mid = (first + last) / 2;
    //     if (isBadVersion(mid)) {
    //         if (!isBadVersion(mid - 1))
    //             return mid;
    //         return checker(first, mid - 1);
    //     }
    //     return checker(mid + 1, last);
    // }
}
